### ISSUE TITLE
net/xrdp-devel on FreeBSD doesn't acknowledge terminated sessions

### ISSUE THREAD

**Issue by derekschrock** [neutrinolabs/xrdp#1016]
Using net/xrdp-devel (0.9.5) on FreeBSD when terminating a session (allowing the window manager to quit) the session is not terminated by sesman and any additional reloads sesman treats them as a reconnected session not a new session.  I was also able reproduce with an update to the port locally using commit e16029c.

```
...
[20180130-20:36:05] [INFO ] Xvnc :10 -auth .Xauthority -geometry 1440x900 -depth 32 -rfbauth /home/dereks/.vnc/sesman_dereks_passwd:10 -bs -nolisten tcp -localhost -dpi 96
...
[20180130-20:36:05] [CORE ] waiting for window manager (pid 3677) to exit
[20180130-20:36:22] [CORE ] window manager (pid 3677) did exit, cleaning up session
[20180130-20:36:22] [INFO ] calling auth_stop_session and auth_end from pid 3676
...
[20180130-20:36:37] [INFO ] A connection received from ::1 port 23800
[20180130-20:36:38] [INFO ] ++ reconnected session: username dereks, display :10.0, session_pid 3676, ip ::ffff:10.20.30.70:53575 - socket: 12
...
```

Later in xrdp.log during the "reconnect":

```
...
[20180130-20:36:38] [DEBUG] return value from xrdp_mm_connect 0
[20180130-20:36:38] [INFO ] xrdp_wm_log_msg: login successful for display 10
[20180130-20:36:38] [DEBUG] xrdp_wm_log_msg: VNC started connecting
[20180130-20:36:38] [DEBUG] xrdp_wm_log_msg: Waiting 2000 ms for VNC to start...
[20180130-20:36:40] [DEBUG] xrdp_wm_log_msg: VNC connecting to 127.0.0.1 5910
[20180130-20:36:43] [DEBUG] VNC error 1 after security negotiation
[20180130-20:36:43] [DEBUG] VNC error before sending share flag
[20180130-20:36:43] [DEBUG] VNC error before receiving server init
[20180130-20:36:43] [DEBUG] VNC error before receiving pixel format
[20180130-20:36:43] [DEBUG] VNC error before receiving name length
[20180130-20:36:43] [DEBUG] VNC error before receiving name
[20180130-20:36:43] [DEBUG] xrdp_wm_log_msg: VNC error - problem connecting
...
```

Restarting the xrdp service fixes this issue however once you terminate a new session the issue reappears.

Is xrdp not getting the SIGCHLD signal that handles terminating and updating the internal state of remembered connections on FreeBSD?

**Comment by derekschrock** [comment]
Also, xrdp appears to not terminate Xorg and X11rdp sessions as well.  However, there are other issues with these that might want to be addressed separately.

**Comment by metalefty** [comment]
Hi, you're a few friends using xrdp on FreeBSD! I'll look at the issue.
In your report, you're using a Xvnc session. Can you confirm if the issue occurs with xorgxrdp?
I think it probably occurs as well as Xvnc.

**Comment by derekschrock** [comment]
With Xorg it just times out and print the following to xrdp.log [DEBUG] Closed socket 19 (AF_UNIX):

```
[20180130-21:30:00] [DEBUG] return value from xrdp_mm_connect 0
[20180130-21:30:00] [INFO ] xrdp_wm_log_msg: login successful for display 41
[20180130-21:30:00] [DEBUG] xrdp_wm_log_msg: started connecting
[20180130-21:30:04] [DEBUG] Closed socket 19 (AF_UNIX)
[20180130-21:30:07] [DEBUG] Closed socket 19 (AF_UNIX)
[20180130-21:30:11] [DEBUG] Closed socket 19 (AF_UNIX)
...
[20180130-21:33:30] [DEBUG] Closed socket 19 (AF_UNIX)
[20180130-21:33:33] [DEBUG] xrdp_wm_log_msg: connection problem, giving up
[20180130-21:33:33] [DEBUG] Closed socket 19 (AF_UNIX)
[20180130-21:33:33] [DEBUG] xrdp_wm_log_msg: some problem
[20180130-21:33:33] [DEBUG] xrdp_mm_module_cleanup
[20180130-21:33:33] [DEBUG] Closed socket 18 (AF_INET6 ::1 port 62692)
[20180130-21:33:33] [DEBUG] Closed socket 12 (AF_INET6 ::ffff:10.20.30.10 port 3389) 
[20180130-21:33:33] [DEBUG] xrdp_mm_module_cleanup
```
Will go on for ~3 minutes until "xrdp_wm_log_msg: connection problem, giving up"

And from xrdp-sesman.log:

```
...
[20180130-21:30:00] [INFO ] ++ created session (access granted): username dereks, ip ::ffff:10.20.30.70:54083 - socket: 12
[20180130-21:30:00] [INFO ] starting Xorg session...
[20180130-21:30:00] [DEBUG] Closed socket 9 (AF_INET6 :: port 5941)
[20180130-21:30:00] [DEBUG] Closed socket 9 (AF_INET6 :: port 6041)
[20180130-21:30:00] [DEBUG] Closed socket 9 (AF_INET6 :: port 6241)
[20180130-21:30:00] [INFO ] calling auth_start_session from pid 8624
[20180130-21:30:00] [DEBUG] Closed socket 8 (AF_INET6 ::1 port 3350)
[20180130-21:30:00] [DEBUG] Closed socket 7 (AF_INET6 ::1 port 3350)
[20180130-21:30:00] [DEBUG] Closed socket 8 (AF_INET6 ::1 port 3350)
[20180130-21:30:00] [INFO ] Xorg :41 -auth .Xauthority -config xrdp/xorg.conf -noreset -nolisten tcp -logfile .xorgxrdp.%s.log
[20180130-21:30:10] [ERROR] X server for display 41 startup timeout
[20180130-21:30:10] [CORE ] waiting for window manager (pid 8625) to exit
[20180130-21:30:10] [ERROR] X server for display 41 startup timeout
[20180130-21:30:10] [ERROR] another Xserver might already be active on display 41 - see log
[20180130-21:30:10] [DEBUG] aborting connection...
[20180130-21:30:10] [CORE ] window manager (pid 8625) did exit, cleaning up session
[20180130-21:30:10] [INFO ] calling auth_stop_session and auth_end from pid 8624
[20180130-21:30:10] [DEBUG] cleanup_sockets:
[20180130-21:30:10] [DEBUG] cleanup_sockets: deleting /tmp/.xrdp/xrdp_chansrv_socket_41
[20180130-21:30:10] [DEBUG] cleanup_sockets: deleting /tmp/.xrdp/xrdpapi_41
[20180130-21:30:10] [DEBUG] cleanup_sockets: failed to delete /tmp/.xrdp/xrdpapi_41
[20180130-21:30:43] [INFO ] shutting down sesman 1
...
```

**Comment by metalefty** [comment]
I can reproduce the issue with xorgxrdp. It occurs with both session type. As you say, sesman's internal state is somehow corrupted.

The issue you can't run xorgxrdp session is another issue so I don't help you with the xorgxrdp launch issue at this time.

**Comment by metalefty** [comment]
sesman deregister the session when session terminated.

when the child sesman (331370) exists, the parent sesman (33168) acknowledge the session termination. parent sesman should receive SIGCHLD but somehow the signal is not received or handled or sent.

```
init(1)─┐
        ├─xrdp(33165)───xrdp(33169)
        └─xrdp-sesman(33168)───xrdp-sesman(33170)─┬─Xorg(33172,meta)
                                                  ├─sh(33171,meta)───xterm(33176)───zsh(33179)
                                                  └─xrdp-chansrv(33178,meta)
```

**Comment by metalefty** [comment]
@derekschrock  Turned out this FreeBSD special code to solve another bug causes this bug.
https://github.com/neutrinolabs/xrdp/blob/08f847d74b8e5b37650746f48c1333453cb04b15/sesman/session.c#L477-L504

I need to brush up the patch.

**Comment by derekschrock** [comment]
I can confirm this patch does fix Xvnc.

**Comment by metalefty** [comment]
Solved by #1030.
